### DISCUSSION TITLE
About Laravel's query builder returning ID list for multiple record registration (PostgreSQL,returning)

### DISCUSSION THREAD

**Discussion by UNILORN** [laravel/framework#51408] [Q&A]
I am using PostgreSQL and considering bulk registration and modification using [Upsert](https://laravel.com/docs/11.x/eloquent#upserts) or [Insert](https://laravel.com/docs/11.x/eloquent#inserts).

In order to get a list of IDs of the added records, I would like to use PostgreSQL's [Returning](https://www.postgresql.org/docs/current/dml-returning.html), but for the query builder, I could only find an implementation for one record.  `insertGetId`

```
psql=> create table t (id serial not null, x varchar not null);
psql=> insert into t (x) values ('a'),('b'),('c') returning id;
 id 
----
  1
  2
  3
(3 rows)
```

https://github.com/laravel/framework/blob/11.x/src/Illuminate/Database/Query/Builder.php#L3651

Is there a reason why the functionality of returning IDs for multiple record inserts is not provided?

If necessary, I am considering contributing a new feature.

**Comment by UNILORN** [comment]
We are also considering creating a PR to create a new function for Postgres. Please let us know if there is an intention to not have these created.

**Comment by hafezdivandari** [comment]
Related to https://github.com/laravel/framework/pull/51014#issuecomment-2119256553

**Comment by TABbl** [comment]
I've opened a PR implementing the base insertReturning case discussed here: [#61286](https://github.com/laravel/framework/pull/61286)

**Comment by wasifahmed85** [comment]
Laravel's query builder doesn't expose RETURNING for bulk inserts because insert() is designed to work across all supported drivers (MySQL, SQLite, SQL Server, etc.), and only PostgreSQL natively supports RETURNING on INSERT. Since the builder has to stay driver-agnostic, insertGetId() was only built for the single-row case (using lastInsertId()/RETURNING id under the hood).

**Reply by TABbl** [reply]
I didn't understand what you meant. Laravel 13.x added insertOrIgnoreReturning for postgres and sqlite, I just suggested adding a regular insert with returning without ignoring conflicts. Because it is often necessary to keep in mind that there was a conflict when inserting records.
